[IV-24-25] Objetivo 2

Mario25402 opened the pull request:
## Sobre la estructura del repositorio

* [x] ¿He seguido las mejores prácticas en el nombre de las clases y ficheros y
   disposición de los mismos?
* [x] ¿He mirado tanto reglas generales (para cualquier repo) como reglas
específicas para el lenguaje con el que se está trabajando?

## Sobre el planteamiento

* [x] ¿Se han planteado una serie de issues, asignados a las historias de
usuario pertinentes, que clarifiquen qué es lo que necesitan los usuarios e
identifiquen las diferentes partes del problema?
* [x] ¿Los issues representan un problema, y no una tarea?

## Sobre el análisis del problema

* [x] ¿Se ha documentado qué análisis se ha hecho sobre el dominio para decir lo
   que se ha creado?
* [x] ¿Se ha documentado por qué se ha elegido que lo creado sea un objeto valor,
   una entidad o un agregado?
* [x] ¿He propuesto un producto mínimamente viable, que en muchos casos será un solo
   objeto valor que no dependa de ningún otro (y que sea la base de muchos
   otros)?.

## Sobre el código

* [x] ¿Todos los mensajes de commit explican el cambio, y no se
      limitan a repetir el nombre del fichero que se ha cambiado?
* [x] ¿Los mensajes de commit siguen el formato estándar y buenas prácticas?
* [x] ¿Se ha hecho una revisión real del código para comprobar que todos
   los atributos y funciones creadas están respaldadas por una HU?
* [x] ¿Todos los issues creados están relacionados con a una HU?
* [x] ¿Ha asignado el propietario/la propietaria del repositorio todos los
issues planteados al *milestone* correspondiente?
* [x] ¿Todos los cambios en el código están asignados a un issue al que se
      referencia en un commit?
* [x] ¿Se ha asignado al mismo milestone el PR que se ha hecho?
* [x] ¿Es el milestone sobre el que estoy trabajando un producto
   mínimamente viable? ¿O tengo que solicitar al product manager que precise de
   qué producto se trata?
* [x] ¿Se ha comprobado que el código entregado sea sintácticamente
correcto?

Mario25402 committed:
Soluciona #7 añadiendo la configuración.

Elección del lenguaje a utilizar.
Entidad del proyecto.

Mario25402 committed:
Corrige #7 cambiando la ruta.

Mario25402 committed:
Closes #8 definiendo la entidad.

Creación de la estructura que contiene la información sobre libro.

- Número de edición es positivo.
- Año de  publicación es positivo.

Mario25402 committed:
Closes #7 con fichero necesario.

Mario25402 committed:
Closes #9 añadiendo el agregado.

Incluye objetos valor y la entidades libro que conforman la bibliografía.

- El año debe ser positivo.

Mario25402 committed:
Closes #10 definiendo el objeto valor.

No es una entidad ya que estos datos no son suficientes para identificar un libro en concreto.

eigenric commented:
Recuerdo que tuvimos un problema en el que se cerró el PR anterior. Ahi yo comenté que había problemas de visibilidad de exportación por estar las iniciales de los métodos y structs en minúscula que veo que se han solucionado.  Igualmente se han simplificado los datos clave de libro tal y como hablamos con el profesor.

Faltaría crear una carpeta model con todos los ficheros pertenecientes al mismo paquete para facilitar la importación en futuro.

```
├── src
  │    └── model
  │         ├── datosLibro.go
  │         ├── guiadocente.go
  │         └── libro.go
  └── go.mod
```

Mario25402 committed:
Corrige #7 cambiando el fichero de ruta

Mario25402 committed:
Corrige #10 cambiando el fichero de ruta

Mario25402 committed:
Corrige #9 cambiando el fichero de ruta

Mario25402 committed:
Corrige #8 cambiando el fichero de ruta

Mario25402 committed:
Corrige #8 añadiendo valores por defecto.

A petición del cliente, se añade que se incluyan valores por defecto en las variables enteras.

eigenric commented:
Considero que se ha entregado el código correcto que define las estructuras a modelar que se pedía en el M0, siguiendo el proceso de división de la H01 en Issues cerrados mediante mensajes de commits atómicos.

* [x] ¿Los issues que ha planteado la persona son efectivamente problemas y
evidencian comprensión de las historias de usuario? Si no es así, es mejor que
se le transmita antes de que empiece a crear cualquier código erróneo.
* [x] ¿Has comprobado que efectivamente el compañero/a ha seguido su lista de
comprobación?
* [x] ¿El código creado es una modelización válida de las historias de usuario
creadas por ti, y te sirve como base para programar la lógica de negocio?
* [x] ¿El código creado sigue las mejores prácticas para el lenguaje de
programación que hayáis consensuado, tanto en identificadores, como
nombres de ficheros, como disposición de los diferentes ficheros en el
repositorio? Por ejemplo, ¿todos los commits se refieren a un issue y
aportan código para resolver el problema planteado en el mismo?
* [x] ¿He comprobado que el código compile, o que quien lo ha programado ha lo
ha hecho y documentado en el repositorio?
* [x] ¿He añadido el PR y todos los issues que se resuelvan en él al milestone
correspondiente, que será el primero?
* [x] ¿Cubre totalmente el milestone 0, o hará falta modelización adicional en
el objetivo 4 para poder programar la lógica de negocio?
* [x] En resumen, ¿considero que con lo hecho puedo ponerme a trabajar
directamente en el siguiente milestone, que incluye la lógica de negocio?

Mario25402 commented:
@JJ

JJ commented:
@eigenric debes aprobar explícitamente.

JJ commented on src/model/libro.go:
```diff
@@ -0,0 +1,67 @@
+}
+
+// Comprueba que un dato sea positivo
+func EsPositivo (num int) bool{
```
Debes seguir las mejores prácticas en los nombres de los argumentos también. `num` no es un nombre suficientemente explicativo.

Mario25402 replied:
Corregido en https://github.com/eigenric/bibliofetch/commit/057912415396b200c1b813aa1ee06db267b34360.

JJ commented on src/model/libro.go:
```diff
@@ -0,0 +1,67 @@
+}
+
+// Establece valores por defecto si no existen (función privada)
+func setDefaults (obj interface{}) {
```
Esto es funcionalidad o lógica de negocio, no es parte del modelo.

Mario25402 replied:
Se ha eliminado ya que se considera una funcionalidad y no se engloba en el M0. https://github.com/eigenric/bibliofetch/commit/c4eb6b6c15d2a92e40a2eea39d5e64fb2264aded

JJ commented on src/model/guiadocente.go:
```diff
@@ -0,0 +1,27 @@
+
+
+// Crea una nueva guía docente con datos válidos
+func NuevaGuiaDocente(anio int, asignatura string, basica []Libro, complementaria []Libro) (*GuiaDocente, error){
```
Los objetos no siempre se crean directamente con los valores de las variables. Para eso no hace falta un objeto. Las clases no solo agrupan, sino que tienen lógica de negocio para crearse a si mismas.
En la HU dice qeu la guía docente está en una página web, es decir, un fichero en este caso. ¿No sería razonable que ese fuera el argumento?

Mario25402 replied:
Se cambia para que sea a partir del descriptor del fichero, desde el cuál se creará el objeto. https://github.com/eigenric/bibliofetch/pull/13/commits/47cdb971ec48130cfe714b35ed142b3359380d7a

JJ reviewed (changes requested):
Mira también los comentarios, por favor.

Mario25402 committed:
Corrige #8 adaptando el argumento.

Se cambia el nombre del argumento para que sea más claro y descriptivo, siguiendo así las recomendaciones del lenguaje.

Mario25402 committed:
Corrige #8 eliminando código del M1.

Se elimina la funcionalidad de añadir valores predeterminados al crear un nuevo libro ya que es una funcionalidad y será añadida en el M1 si se necesita.

Mario25402 committed:
Corrige #9 cambiando el constructor.

Se cambia la lógica del constructor para que funcione a partir de los datos de origen. Se hace desde un descriptor ya que el propia función procesa datos y no deben dedicarse a diferentes funcionalidades.

Mario25402 committed:
Borra código antiguo de #9.

Mario25402 committed:
Elimina los comentarios de #9.

Mario25402 committed:
Elimina contenido innecesario de #8

Mario25402 commented:
Sugerencias corregidas @JJ

JJ reviewed (approved):
:+1:
